multicluster/controllers/multicluster/memberclusterannounce_controller.go

```diff
@@ -287,3 +322,67 @@ func (r *MemberClusterAnnounceReconciler) GetMemberClusterStatuses() []multiclus
 
 	return status
 }
+
+func (r *MemberClusterAnnounceReconciler) addMemberToClusterSet(memberClusterAnnounce *multiclusterv1alpha1.MemberClusterAnnounce) error {
+	clusterSetID := memberClusterAnnounce.ClusterSetID
+	clusterSet := &multiclusterv1alpha1.ClusterSet{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Namespace: memberClusterAnnounce.Namespace, Name: clusterSetID}, clusterSet); err != nil {
+		if errors.IsNotFound(err) {
+			klog.ErrorS(err, "ClusterSet not found in leader cluster", "ClusterSet", clusterSetID)
+			return err
+		}
+		klog.ErrorS(err, "Failed to get ClusterSet in leader cluster", "ClusterSet", clusterSetID)
+		return err
+	}
+
+	exist := false
+	for _, member := range clusterSet.Spec.Members {
+		if member.ClusterID == memberClusterAnnounce.ClusterID {
+			exist = true
+			break
+		}
+	}
+
+	if exist {
+		return fmt.Errorf("member cluster %s already exists in ClusterSet %s", memberClusterAnnounce.ClusterID, clusterSetID)
+	}
+
+	clusterSet.Spec.Members = append(clusterSet.Spec.Members, multiclusterv1alpha1.MemberCluster{ClusterID: memberClusterAnnounce.ClusterID})
+	if err := r.Update(context.TODO(), clusterSet); err != nil {
+		klog.ErrorS(err, "Failed to update ClusterSet in leader cluster", "ClusterSet", clusterSetID, "Namespace", clusterSet.Namespace)
+		return err
+	}
+	return nil
+}
+
+func (r *MemberClusterAnnounceReconciler) removeMemberFromClusterSet(memberClusterAnnounce *multiclusterv1alpha1.MemberClusterAnnounce) error {
+	clusterSetID := memberClusterAnnounce.ClusterSetID
+	clusterSet := &multiclusterv1alpha1.ClusterSet{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Namespace: memberClusterAnnounce.Namespace, Name: clusterSetID}, clusterSet); err != nil {
+		if errors.IsNotFound(err) {
+			klog.ErrorS(err, "ClusterSet not found in leader cluster", "ClusterSet", clusterSetID)
+			return err
+		}
+		klog.ErrorS(err, "Failed to get ClusterSet in leader cluster", "ClusterSet", clusterSetID)
+		return err
+	}
+
+	found := false
+	for i, member := range clusterSet.Spec.Members {
+		if member.ClusterID == memberClusterAnnounce.ClusterID {
+			found = true
+			clusterSet.Spec.Members = append(clusterSet.Spec.Members[:i], clusterSet.Spec.Members[i+1:]...)
+			break
+		}
+	}
+	if !found {
+		klog.InfoS("Member cluster not found in ClusterSet", "ClusterSet", clusterSetID, "Member", memberClusterAnnounce.ClusterID)
+		return fmt.Errorf("member cluster %s not found in ClusterSet %s", memberClusterAnnounce.ClusterID, clusterSetID)
```

Comment:
This error should also be ignored? maybe just return immediately if found is false. Otherwise, reconciler will retry again and again when there is an error returned.

Reply:
Sure. NotFound error ignored.